### PR TITLE
[rest-server] add user vc management and virtual cluster access control

### PR DESCRIPTION
**Pull request by wangcan0329**

### CODE REVIEW

File: rest-server/test/jobSubmission.js [hunk 1]
```diff
@@ -4,7 +4,7 @@
 // MIT License
 //
 // Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
-// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+// documentation files (the 'Software'), to deal in the Software without restriction, including without limitation
```
Comment (abuccts) [inline rest-server/test/jobSubmission.js]: do not need to change?

File: rest-server/test/jobSubmission.js [hunk 2]
```diff
@@ -154,7 +215,7 @@ describe('Submit job: POST /api/v1/jobs', () => {
     global.chai.request(global.server)
       .post('/api/v1/jobs')
       .set('Authorization', 'Bearer ' + validToken)
-      .send(JSON.parse(global.mustache.render(global.jobConfigTemplate, {'jobName': 'job1'})))
+      .send(JSON.parse(global.mustache.render(global.jobConfigTemplate, { 'jobName': 'job1' })))
```
Comment (abuccts) [inline rest-server/test/jobSubmission.js]: `object-curly-spacing` lint error, e.g., in [build #1125](https://travis-ci.org/Microsoft/pai/jobs/369736701)

File: rest-server/test/jobSubmission.js (hunk 1 above)
Comment (wangcan0329) [reply rest-server/test/jobSubmission.js]: Sorry, I replace " to ' by mistake. I will revert this change

File: rest-server/src/models/user.js [hunk 3]
```diff
@@ -133,6 +134,93 @@ const remove = (username, callback) => {
...
+            if (!updateVcList.includes('default')) { // always has 'default' queue
+              updateVcList.push('default');
+            }
+            updateVcList.sort();
+            db.set(etcdConfig.userVirtualClusterPath(username), updateVcList.toString(), null, (errMsg, res) => {
+              if (errMsg) {
+                logger.warn('update %s virtual cluster: %s failed, error message:%s', etcdConfig.userVirtualClusterPath(username), errMsg);
+                callback(errMsg, false);
+              } else {
+                if (addUserWithInvalidVc) {
```
Comment (Asakuri) [inline rest-server/src/models/user.js]: why use closure here and why put this in db.set callback?

Comment (wangcan0329) [reply rest-server/src/models/user.js]: When add new user if input virtual cluster is invalid we also need to add this user to default group, and we also should let user know the input is invalid.

File: rest-server/src/models/user.js [hunk 4]
```diff
@@ -133,6 +134,93 @@ const remove = (username, callback) => {
...
+  } else {
+    db.get(etcdConfig.userPath(username), null, (errMsg, res) => {
+      if (errMsg) {
+        logger.warn('user %s not exists', etcdConfig.userPath(username));
+        callback(errMsg, false);
+      } else {
+        VirtualCluster.prototype.getVcList((vcList, err) => {
+          if (err) {
+            logger.warn('get virtual cluster list error\n%s', err.stack);
+          } else if (vcList === undefined) {
```
Comment (hwuu) [inline rest-server/src/models/user.js]: else if (!vcList) ...

File: rest-server/src/models/user.js (hunk 3 above)
Comment (hwuu) [reply rest-server/src/models/user.js]: @asakuri: I discussed with Can yesterday. Her philosophy is to add user with best efforts. Do you have some examples of pros and cons of rejecting the whole request? 

And @can: I suggest you include the VCs that are not valid in the http response, we should not do anything in the background out of admin’s expectation.

File: rest-server/src/controllers/job.js [hunk 5]
```diff
@@ -115,10 +115,22 @@ const update = (req, res) => {
   Job.prototype.putJob(req.job.name, req.body, req.user.username, (err) => {
     if (err) {
       logger.warn('update job %s error\n%s', req.job.name, err.stack);
-      return res.status(500).json({
-        error: 'JobUpdateError',
-        message: 'job update error',
-      });
+      if (err.message === 'VirtualClusterNotFound') {
```
[the remark this answers is not in the document]
Comment (wangcan0329) [reply rest-server/src/controllers/job.js]: discussed with Hao, we will define all error handling later

File: rest-server/src/models/user.js [hunk 6]
```diff
@@ -133,6 +134,93 @@ const remove = (username, callback) => {
...
+const updateUserVc = (username, virtualClusters, callback) => {
+  if (typeof username === 'undefined') {
+    callback(new Error('user does not exist'), false);
+  } else {
+    db.get(etcdConfig.userPath(username), null, (errMsg, res) => {
+      if (errMsg) {
+        logger.warn('user %s not exists', etcdConfig.userPath(username));
+        callback(errMsg, false);
+      } else {
+        VirtualCluster.prototype.getVcList((vcList, err) => {
```
[the remark this answers is not in the document]
Comment (wangcan0329) [reply rest-server/src/models/user.js]: I call getVcList api which defines vcList as first parameter

File: rest-server/src/models/user.js [hunk 7]
```diff
@@ -133,6 +134,93 @@ const remove = (username, callback) => {
   }
 };
 
+const updateUserVc = (username, virtualClusters, callback) => {
+  if (typeof username === 'undefined') {
+    callback(new Error('user does not exist'), false);
+  } else {
+    db.get(etcdConfig.userPath(username), null, (errMsg, res) => {
```
[the remark this answers is not in the document]
Comment (wangcan0329) [reply rest-server/src/models/user.js]: Discussed with Hao, we will define all rest-server error handling later

File: rest-server/src/routes/user.js [hunk 8]
```diff
@@ -31,5 +31,8 @@ router.route('/')
     /** DELETE /api/v1/user - Remove a user */
     .delete(tokenConfig.check, param.validate(userConfig.userDeleteInputSchema), userController.remove);
 
+router.route('/:username/virtualClusters')
+    .put(tokenConfig.check, param.validate(userConfig.userVcUpdateInputSchema), userController.updateUserVc);
```
Comment (hwuu) [inline rest-server/src/routes/user.js]: There is no 'GET' API?

Comment (wangcan0329) [reply rest-server/src/routes/user.js]: we only add access control part, not support 'GET'

Comment (wangcan0329) [reply rest-server/src/routes/user.js]: Only add access control part, we don't support 'GET' API now.

**Pull request merged**